launcher/src/main/java/org/apache/spark/launcher/AbstractAppHandle.java

```diff
@@ -29,15 +30,15 @@
 
   private final LauncherServer server;
 
-  private LauncherConnection connection;
+  private LauncherServer.ServerConnection connection;
   private List<Listener> listeners;
-  private State state;
+  private AtomicReference<State> state;
```

Comment:
What's the rationale behind this? Previously we just make sure all access and modification to state is synchronized, are you changing it to `AtomicReference` for performance reasons?

Reply:
With the new code, synchronization would cause a deadlock.

- handle calls `closeAndWait()` inside synchronized block which joins connection thread
- connection thread would call `setState()` on the handle and cause a deadlock

Changing the state should be as thread-safe as before with the new code.

Reply:
makes sense